src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -6542,7 +6543,13 @@ IArrayInitializerOperation popAndAssembleArrayInitializerValues(IArrayInitialize
         public override IOperation? VisitCollectionExpression(ICollectionExpressionOperation operation, int? argument)
         {
             EvalStackFrame frame = PushStackFrame();
-            var creationArguments = VisitArguments(operation.ConstructArguments, instancePushed: false);
+
+            // Ugly, but necessary.  If we bound successfully, we'll have an array of IArgumentOperation.  We want to
+            // call through to VisitArguments to handle it properly.  So attempt to cast to that type first, but
+            // fallback to just visiting the array of expressions if we didn't bind successfully.
+            var creationArguments = ImmutableCollectionsMarshal.AsArray(operation.ConstructArguments) is IArgumentOperation[] argumentArray
```

Comment:
It looks like [`ImmutableArray<T>.As<TOther>()`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray-1.as?view=net-9.0#remarks) might work here. e.g.

`operation.ConstructArguments.As<IArgumentOperation>() is { IsDefault: false } argumentArray ? ...`

Reply:
much nicer!